core/templates/dev/head/pages/admin/roles_tab/RolesTabDirective.js

```diff
@@ -0,0 +1,129 @@
+// Copyright 2017 The Oppia Authors. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS-IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+/**
+ * @fileoverview Directive for the Roles tab in the admin panel.
+ */
+
+oppia.directive('adminRolesTab', [
+  '$http', 'ADMIN_ROLE_HANDLER_URL', 'AdminTaskManagerService',
+  'UrlInterpolationService', 'ADMIN_SHOW_UPDATE_ROLE',
+  function(
+    $http, ADMIN_ROLE_HANDLER_URL, AdminTaskManagerService,
+    UrlInterpolationService, ADMIN_SHOW_UPDATE_ROLE) {
+    return {
+      restrict: 'E',
+      scope: {
+        setStatusMessage: '='
+      },
+      templateUrl: UrlInterpolationService.getDirectiveTemplateUrl(
+        '/pages/admin/roles_tab/roles_tab_directive.html'),
+      controller: ['$scope', function($scope) {
+        $scope.UPDATABLE_ROLES = GLOBALS.UPDATABLE_ROLES;
+        $scope.VIEWABLE_ROLES = GLOBALS.VIEWABLE_ROLES;
+        $scope.graphData = GLOBALS.ROLE_GRAPH_DATA;
+        $scope.showResultRoles = false;
+        $scope.result = {};
+        $scope.setStatusMessage('');
+        $scope.showUpdateForm = ADMIN_SHOW_UPDATE_ROLE;
+
+        // Calculating initStateId and finalStateIds for graphData
+        // Since role graph is acyclic, node with no incoming edge
+        // is initState and nodes with no outgoing edge are finalStates.
+        var hasIncomingEdge = [];
+        var hasOutgoingEdge = [];
+        for (var i = 0; i < $scope.graphData.links.length; i++) {
+          hasIncomingEdge.push($scope.graphData.links[i].target);
+          hasOutgoingEdge.push($scope.graphData.links[i].source);
+        }
+        var finalStateIds = [];
+        for (var i in $scope.graphData.nodes) {
```

Comment:
can you use a more descriptive variable name than `i`? I believe they're supposed to be role IDs here?